### PR TITLE
Added possibility to set force_id when post an object

### LINKED ISSUE #19847: Add possibility to specify id inserting object from webService
#### Is your feature request related to a problem?

I am adding to our shop management desktop program to prestashop eCommerce, and I want use our id for categories and products.

#### Describe the solution you'd like

I would like a web interface similar to CSV upload, in that case you can specify the id, and the element are added or updated.

#### Alternatives you've considered

Another possibility is specify the id on post or add the id_backoffice or id_desktop or another name.

#### Additional context

none

### PR DESCRIPTION
**Pull request by APerricone**
| Questions     | Answers
| ------------- | -------------------------------------------------------
| Branch?       | develop
| Description?  | Added possibilty to force_id when using webService, it is already possibile using php classes
| Type?         | improvement
| Category?     | WS
| BC breaks?    | no
| Deprecations? | no
| Fixed ticket? |  #19847
| How to test?  | Do some posts specifing id with force_id = 0 and 1, with multiple entitities too

---
This change is [Reviewable](https://reviewable.io/reviews/prestashop/prestashop/19841)

### CODE REVIEW

**Comment by NeOMakinG** [comment]
Hey @APerricone, can you create an issue ? It's needed for the changelog, also the team may discuss about this feature ! 

Thanks for your contribution anyway :)

**Comment by APerricone** [comment]
just create the issue #19847 with the context of my problem.

I created a post on forum https://www.prestashop.com/forums/topic/1025782-save-id-from-another-software/ but it needs to be approved

File: classes/webservice/WebserviceRequest.php [hunk 1]
```diff
@@ -1510,11 +1510,15 @@ protected function saveEntityFromXml($successReturnCode)
         $object = null;
 
         $ids = [];
+        $ids2 = [];
```
Comment (Progi1984) [inline classes/webservice/WebserviceRequest.php]:
```suggestion
        $forceIds = [];
```

File: classes/webservice/WebserviceRequest.php [hunk 2]
```diff
@@ -1510,11 +1510,15 @@ protected function saveEntityFromXml($successReturnCode)
...
         foreach ($xmlEntities as $entity) {
             // To cast in string allow to check null values
             if ((string) $entity->id != '') {
                 $ids[] = (int) $entity->id;
             }
+            if ((int) $entity->force_id > 0) {
+                $ids2[] = (int) $entity->force_id;
```
Comment (Progi1984) [inline classes/webservice/WebserviceRequest.php]:
```suggestion
                $forceIds[] = (int) $entity->force_id;
```

File: classes/webservice/WebserviceRequest.php [hunk 3]
```diff
@@ -1530,9 +1534,11 @@ protected function saveEntityFromXml($successReturnCode)
                 return false;
             }
         } elseif ($this->method == 'POST' && count($ids) > 0) {
-            $this->setError(400, 'id is forbidden when adding a new resource', 91);
+            if (count($ids2) < count($ids)) {
```
Comment (Progi1984) [inline classes/webservice/WebserviceRequest.php]:
```suggestion
            if (count($forceIds) < count($ids)) {
```

**Comment by NeOMakinG** [comment]
Thanks a lot @APerricone

File: classes/webservice/WebserviceRequest.php [hunk 4]
```diff
@@ -1542,7 +1548,23 @@ protected function saveEntityFromXml($successReturnCode)
...
+                    $object = new $this->resourceConfiguration['retrieveData']['className']((int) $attributes->id);
+                    if ($object->id) {
+                        $this->setError(400, 'id is already used', 92);
+
+                        return false;
+                    }
+                    if ((int) $attributes->force_id > 0) {
+                        $object->force_id = true;
+                    } else {
+                        $this->setError(400, 'id is forbidden when adding a new resource', 91);
```
Comment (APerricone) [inline classes/webservice/WebserviceRequest.php:1561]: with the check on line 1537 it should never happen.
Sorry, it was written before.

**Comment by matks** [comment]
I am afraid of this change.
I'm sorry @APerricone your PR seems fine but the PrestaShop webservice is a overly complex piece of softwate built on a very fragile concept and there is no internal documentation nor tests available to allow us to modify it without fear of breaking anything.

Basically we have no reliable way to check your PR does not break anything, and your PR brings quite an impactul change. Modifying how IDs are generated can have huge consequences on a system such as PrestaShop which has a huge of business logic encoded into SQL.

I provided some ideas for workarounds in the issue you created but I'm very wary of merging this PR

**Comment by matks** [comment]
Hi everone

It looks like this PR is not moving anymore. As mentioned in our [Maintainer Guide](https://devdocs.prestashop.com/1.7/project/maintainers-guide/reviewing-pull-requests/#stale-pull-requests), "Pull Requests may be closed after 30 days of inactivity following a request for modifications".

I close this pull request as it seems stale and you might dont want to continue on it. But I'm open for the discussion 😉  if we could find items to validate this PR is not risky, we could merge it. Please tell me 😉  .

Thank you!

**Pull request closed without merging**